### ISSUE TITLE
z_stream is must be pinned after being initialized

### ISSUE THREAD

**Issue by taralx** [rust-lang/libz-sys#59]
inflateInit allocates a `state` object that points back to the `z_stream`. If it moves after that, then all subsequent calls result in `Z_STREAM_ERROR` due to pointer mismatch. So everyone who is using `z_stream` by value is unfortunately doing it wrong.

**Comment by taralx** [comment]
At minimum, please `impl !Unpin for z_stream`.

**Comment by mattfs** [comment]
I hit this same bug a ways back before I fully understood Pin. +1 to `impl !Unpin`.

**Comment by joshtriplett** [comment]
cc @stjepang

Could you weigh in on this? This makes sense to me, but I would appreciate some review from someone I know has a deep understanding of `Pin` and `Unpin`.

**Comment by stjepang** [comment]
Implementing `!Unpin` for `z_stream` seems like a good idea (btw, `PhantomPinned` marker can help with that). However, that's not very helpful unless functions taking a pointer to `z_stream` are changed to take a `Pin<&mut z_stream>` instead.

Take the following function as an example:

```rust
pub fn deflate(strm: z_streamp, flush: c_int) -> c_int;

pub type z_streamp = *mut z_stream;
```

If we make `z_stream` an `!Unpin` type, then nothing prevents us from misusing it yet. We'd have to change the signature of the function as follows:

```rust
pub fn deflate(strm: Pin<&mut z_stream>, flush: c_int) -> c_int;
```

Having a `Pin<&mut z_stream>` is basically a promise that the `z_stream` behind this pointer will never be moved to another address - and in the end of its life, it will be dropped in-place at the same address (or it won't die and will leak forever).

But, given that this is a "-sys" crate (a direct translation of the C API) and not a high-level API, I don't know if using `Pin` here actually makes sense. Where is this crate used? Is there a higher level API for it?

**Comment by joshtriplett** [comment]
@stjepang The most common use of `libz-sys` is in `flate2`, which provides a higher-level API atop it.

The functions here do indeed take raw pointers, and there's an assumption that once you initialize, you won't move the objects around.

Given that these functions take raw pointers, and that they're an ABI that we shouldn't change, I don't think we should change them to take a `Pin`. That's something for a higher-level safe API to deal with.

Given that, I'm going to close this.

**Comment by taralx** [comment]
Hi, can we at least make it !Unpin so safe wrappers do the right thing? That shouldn't change anything unless people are using Pin already, which is basically no-one.

**Comment by taralx** [comment]
It's not possible to do outside of this crate.

**Comment by dpereira** [comment]
Hi, I've stumbled upon this while using the crate to implement my own Read'ers for encoding / decoding gzip, as the ones in flate2 did not suit my particular use case.

I was able to workaround the issue by allocating the zstream object in the heap by Box'ing it, but then it is needed to pass in the stream mut ref by dereferrencing the box: `&mut *stream`. Just recording here in case this workaround may come handy to someone else.
